src/lib/transaction_logic/test/account_timing.ml

```diff
@@ -0,0 +1,194 @@
+open Core_kernel
+open Mina_transaction_logic
+open Currency
+open Mina_base
+open Mina_numbers
+
+type timing = (Global_slot.t, Balance.t, Amount.t) Account_timing.tt
+[@@deriving compare, sexp]
+
+type txn_result = (timing, Error.t) Result.t [@@deriving compare, sexp]
+
+let final_vesting_slot ~initial_minimum_balance ~cliff_time ~cliff_amount
+    ~vesting_period ~vesting_increment =
+  let open Unsigned in
+  let to_vest =
+    Balance.(initial_minimum_balance - cliff_amount)
+    |> Option.value_map ~default:UInt64.zero ~f:Balance.to_uint64
+  in
+  let incr = Amount.to_uint64 vesting_increment in
+  let periods =
+    let open UInt64 in
+    if compare (rem to_vest incr) zero = 0 then div to_vest incr
+    else succ @@ div to_vest incr
+  in
+  let open UInt32 in
+  let open Infix in
+  Global_slot.to_uint32 cliff_time
+  + (UInt64.to_uint32 periods * Global_slot.to_uint32 vesting_period)
+  |> Global_slot.of_uint32
+
+let timing_final_vesting_slot = function
+  | Account.Timing.Untimed ->
+      Global_slot.zero
+  | Timed
+      { initial_minimum_balance
+      ; cliff_time
+      ; cliff_amount
+      ; vesting_period
+      ; vesting_increment
+      } ->
+      final_vesting_slot ~initial_minimum_balance ~cliff_time ~cliff_amount
+        ~vesting_period ~vesting_increment
+
+let%test_module "Test account timing." =
+  ( module struct
+    (* Test that the timed account generator only generates accounts that
+       are completely vested before the largest possible slot, (2^32)-1. *)
+    let%test_unit "Test fine-tuning of the account generation." =
+      Quickcheck.test Account.gen_timed ~f:(fun account ->
+          [%test_pred: Balance.t]
```

Comment:
I think it would be clearer to use `%test_pred` only on the timed case

Reply:
Agreed. See 32146d18f4.